File: python/4_median_of_two_sorted_arrays.py
# 4. Median of Two Sorted Arrays
#
# There are two sorted arrays nums1 and nums2 of size m and n respectively.
# Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).
#

class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        nums = nums1 + nums2
        nums.sort()
        if len(nums) % 2 == 0:
            return (nums[len(nums) / 2] + nums[len(nums) / 2 - 1]) / 2.0
        else:
            return nums[len(nums) / 2]


def main():
    solution = Solution()
    print(solution.findMedianSortedArrays([], [2, 3]))  # 2.5


if __name__ == '__main__':
    main()

File: go/src/main/4_median_of_two_sorted_arrays.go
package main

import "fmt"

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	nums := make([]int, len(nums1)+len(nums2))
	idx1 := 0
	idx2 := 0
	idx := 0

	for idx1 < len(nums1) && idx2 < len(nums2) {
		if nums1[idx1] < nums2[idx2] {
			nums[idx] = nums1[idx1]
			idx1++
		} else {
			nums[idx] = nums2[idx2]
			idx2++
		}
		idx++
	}

	for idx1 < len(nums1) {
		nums[idx] = nums1[idx1]
		idx1++
		idx++
	}

	for idx2 < len(nums2) {
		nums[idx] = nums2[idx2]
		idx2++
		idx++
	}

	if len(nums)%2 == 0 {
		return float64(nums[len(nums)/2]+nums[len(nums)/2-1]) / 2
	} else {
		return float64(nums[len(nums)/2])
	}
}

func main() {
	for _, unit := range []struct {
		nums1  []int
		nums2  []int
		median float64
	}{
		{[]int{1, 3}, []int{2}, 2.0},
		{[]int{1, 2}, []int{3, 4}, 2.5},
	} {
		if median := findMedianSortedArrays(unit.nums1, unit.nums2); median != unit.median {
			fmt.Printf("findMedianSortedArrays(%v, %v) = %v != %v\n", unit.nums1, unit.nums2, median, unit.median)
		}
	}
}
